### PR TITLE
Fix infinite recursion when __repr__ is called

### PR DESCRIPTION
**Pull request by aucampia**
Also add some more tests.

### CODE REVIEW

**Comment by edmondchuc** [comment]
Thanks for this @aucampia

File: rdflib/namespace.py [hunk 1]
```diff
@@ -141,7 +141,7 @@ def __getattr__(self, name):
         return self.term(name)
 
     def __repr__(self):
-        return f"Namespace({self!r})"
+        return f"Namespace({self})"
```
Comment (kernc) [inline rdflib/namespace.py]: Can we make this:
```suggestion
        return f"Namespace({super().__repr__()})"
```
Or some other way to ensure the output URI is _quoted_ as a [Python expression](https://docs.python.org/3/reference/datamodel.html#object.__repr__).

File: rdflib/namespace.py [hunk 2]
```diff
@@ -225,7 +226,7 @@ def __getattr__(self, name):
                 raise AttributeError(e)
 
     def __repr__(self):
-        return f"{self.__module__}.{self.__class__.__name__}({self!r})"
+        return f"{self.__module__}.{self.__class__.__name__}({str(self)!r})"
```
Comment (aucampia) [inline rdflib/namespace.py:228]: `super().__repr__()` works but results in the rather unsightly:

```python
rdflib.namespace.ClosedNamespace(Namespace('http://example.com/name/space/'))
```

And `super(str).__repr__()` results in:

```python
rdflib.namespace.ClosedNamespace(<super: <class 'str'>, NULL>)
```

File: rdflib/namespace.py (hunk 1 above)
Comment (aucampia) [reply rdflib/namespace.py]: fixed and added tests for it.

File: rdflib/namespace.py [hunk 3]
```diff
@@ -199,6 +199,7 @@ class ClosedNamespace(Namespace):
 
     def __new__(cls, uri, terms):
         rt = super().__new__(cls, uri)
+        rt.__terms = terms
```
Comment (kernc) [inline rdflib/namespace.py]: Is this used?
```suggestion
```

Comment (aucampia) [reply rdflib/namespace.py]: Apologies, I was thinking of fixing __repr__ to show terms, but then decided against that, removed it.

Comment (kernc) [reply rdflib/namespace.py]: Interesting. The `terms` would be needed to reconstruct the `ClosedNamespace` object, but the full list might indeed be unruly for printing to stdout. :thinking:

Comment (aucampia) [reply rdflib/namespace.py]: I think it is best to not try and fix all problems there in one PR, better to get it in to the release and make an issue for the other issues. This is not a problem introduced in your PR.

**Review by kernc — APPROVED** [review]

**Pull request merged**